packages/nbt/src/checker/index.ts

```diff
@@ -286,18 +279,149 @@ export function blockStates(
+		// TODO: support dispatcher
+		const definition = getRegistryIdentifier(registry)
+		if (!definition) {
+			return
+		}
+		// Create a linked list representation
+		const leaf = {
+			type: 'leaf',
+			range: core.Range.create(node.range.end),
+		} as const
+		let link: NbtPathLink = { path: node, node: leaf }
+		for (let i = node.children.length - 1; i >= 0; i -= 1) {
+			link = {
+				path: node,
+				node: node.children[i],
+				next: link,
+			}
+		}
+		let prev = link
+		while (prev.next) {
+			prev.next.prev = prev
+			prev = prev.next
+		}
+		mcdoc.runtime.checker.reference<NbtPathLink>(
+			[{ originalNode: link, inferredType: inferPath(link) }],
+			definition,
+			{
+				context: ctx,
+				isEquivalent: (inferred, def) => {
+					switch (inferred.kind) {
+						case 'list':
+						case 'byte_array':
+						case 'int_array':
+						case 'long_array':
+							return [
+								'list',
+								'tuple',
+								'byte_array',
+								'int_array',
+								'long_array',
+							].includes(def.kind)
+						default:
+							return false
+					}
+				},
+				getChildren: (
+					link,
+				): mcdoc.runtime.checker.RuntimeUnion<NbtPathLink>[] => {
+					while (
+						link.next && link.node.type !== 'leaf' &&
+						NbtCompoundNode.is(link.node)
+					) {
+						link = link.next
+					}
+					if (!link.next || link.node.type === 'leaf') {
+						return []
+					}
+					if (NbtPathIndexNode.is(link.node)) {
+						return [[{
+							originalNode: link.next,
+							inferredType: inferPath(link.next),
+						}]]
+					}
+					if (core.StringNode.is(link.node)) {
+						return [{
+							key: {
+								originalNode: link,
+								inferredType: {
+									kind: 'literal',
+									value: { kind: 'string', value: link.node.value },
+								},
+							},
+							possibleValues: [{
+								originalNode: link.next,
+								inferredType: inferPath(link.next),
+							}],
+						}]
+					}
+					// Never reachable
+					return []
+				},
+				reportError: (error) => {
+					if (
+						error.kind === 'missing_key' ||
+						error.kind === 'invalid_collection_length'
+					) {
+						return
+					}
+					mcdoc.runtime.checker
+						.getDefaultErrorReporter<NbtPathLink>(
+							ctx,
+							({ originalNode: link }) => link.node.range,
+						)(error)
+				},
+				attachTypeInfo: (link, definition) => {
+					// TODO: attach type def
+					// TODO: improve hover info
+					if (core.StringNode.is(link.prev?.node)) {
+						link.prev.node.hover =
+							`\`\`\`typescript\n${link.prev.node.value}: ${
+								mcdoc.McdocType.toString(definition)
+							}\n\`\`\``
+					}
+				},
+				stringAttacher: (node, attacher) => {
+					if (!core.StringNode.is(node)) return
+					attacher(node)
+					if (node.children) {
+						core.AstNode.setParents(node)
+						// Because the runtime checker happens after binding, we need to manually call this
+						core.binder.dispatchSync(node, ctx)
+						core.checker.dispatchSync(node, ctx)
+					}
+				},
```

Comment:
Would it make sense to add a general getDefaultTypeInfoAttacher and getDefaultStringAttacher? The code here looks pretty similar across different consumers.

Reply:
Agreed, this could use a DRY-refactoring